pkg/payerreport/workers/attestation.go

```diff
@@ -0,0 +1,180 @@
+// and signing them with the node's private key.
+type attestationWorker struct {
+	ctx          context.Context
+	cancel       context.CancelFunc
+	log          *zap.Logger
+	registrant   registrant.IRegistrant
+	store        payerreport.IPayerReportStore
+	verifier     payerreport.PayerReportVerifier
+	wg           sync.WaitGroup
+	pollInterval time.Duration
+}
+
+// NewAttestationWorker creates and starts a new attestation worker that will periodically
+// check for reports that need attestation.
+// It takes a context, logger, registrant for signing, store for accessing reports,
+// and a poll interval that determines how often to check for reports.
+func NewAttestationWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	registrant registrant.IRegistrant,
+	verifier payerreport.PayerReportVerifier,
+	store payerreport.IPayerReportStore,
+	pollInterval time.Duration,
+) *attestationWorker {
+	ctx, cancel := context.WithCancel(ctx)
+
+	worker := &attestationWorker{
+		ctx:          ctx,
+		log:          log.Named("attestationworker"),
+		registrant:   registrant,
+		store:        store,
+		verifier:     verifier,
+		wg:           sync.WaitGroup{},
+		cancel:       cancel,
+		pollInterval: pollInterval,
+	}
+
+	worker.start()
+
+	return worker
+}
+
+// start launches the worker's main loop in a separate goroutine.
+// The loop periodically checks for reports that need attestation.
+func (w *attestationWorker) start() {
+	tracing.GoPanicWrap(
+		w.ctx,
+		&w.wg,
+		"attestation-worker",
+		func(ctx context.Context) {
+			w.log.Info("Starting attestation worker")
+			var err error
+			ticker := time.NewTicker(w.pollInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					if err = w.attestReports(); err != nil {
+						w.log.Error("attesting reports", zap.Error(err))
+					}
+
+				}
+			}
+		},
+	)
+}
+
+// attestReports fetches all reports with pending attestation status
+// and attempts to attest each one.
+// Returns an error if there was a problem fetching the reports.
+func (w *attestationWorker) attestReports() error {
+	uncheckedReports, err := w.findReportsNeedingAttestation()
+	if err != nil {
+		return err
+	}
+
+	for _, report := range uncheckedReports {
+		if err := w.attestReport(report); err != nil {
+			w.log.Error("attesting report", zap.Error(err))
+		}
+	}
+
+	return err
+}
+
+func (w *attestationWorker) findReportsNeedingAttestation() ([]*payerreport.PayerReportWithStatus, error) {
+	return w.store.FetchReports(
+		w.ctx,
+		payerreport.NewFetchReportsQuery().WithAttestationStatus(payerreport.AttestationPending),
+	)
+}
+
+// attestReport validates a single report by checking its consistency with previous reports
+// and, if valid, signs it with the node's private key.
+// Returns an error if the report is invalid or if there was a problem during attestation.
+func (w *attestationWorker) attestReport(report *payerreport.PayerReportWithStatus) error {
```

Comment:
for my own clarity, can you walk me through the storage here? Where do we store previous attestations? Does the node need to store the entire history since 1?

Reply:
> Where do we store previous attestations?

We have a table for reports and a table for attestations

> Does the node need to store the entire history since 1?

For the correctness of the system we only need to store the most recent submitted report for each originator, and any unsubmitted reports with a `startSequenceID` > the `endSequenceID` of that report which haven't expired.

Right now I am storing them all, but we can add cleanup after we start moving reports to a submitted status.